src/wallet/walletmanager.h

```diff
@@ -0,0 +1,56 @@
+// Copyright (c) 2018 The Bitcoin Core developers
+// Distributed under the MIT software license, see the accompanying
+// file COPYING or http://www.opensource.org/licenses/mit-license.php.
+
+#ifndef BITCOIN_WALLET_WALLETMANAGER_H
+#define BITCOIN_WALLET_WALLETMANAGER_H
+
+#include <wallet/wallet.h>
+
+#include <vector>
+#include <util.h>
+
+class CWalletManager {
+private:
+    std::vector<CWallet*> m_vpwallets;
+
+public:
+    //! Load wallet databases.
+    bool OpenWallets();
+
+    //! Complete startup of wallets.
+    void StartWallets(CScheduler& scheduler);
+
+    //! Flush all wallets in preparation for shutdown.
+    void FlushWallets();
+
+    //! Stop all wallets. Wallets will be flushed first.
+    void StopWallets();
+
+    //! Close all wallets.
+    void CloseWallets();
+
+    void AddWallet(CWallet *wallet);
+    void DeallocWallet(unsigned int pos);
+
+    bool HasWallets();
+
+    unsigned int CountWallets();
+
+    template<typename Callable>
+    void ForEachWallet(Callable&& func)
```

Comment:
Why not just return a copy of the array?

Reply:
I think a lambda is more flexible... same reason we use it in `CNode`/`g_conman`.

Reply:
In `g_connman` makes sense because `cs_vNodes` is held all the time. Here we don't have locks. Even when we add dynamic wallets, IMHO the same pattern doesn't apply here.

Reply:
IMO, once we add the locking (not the scope of this PR), then it makes more sense to keep the locking inside of the manager instead of returning a copy.